### PR TITLE
[PoemBot][HOC] Validation Code

### PR DESCRIPTION
**Pull request by ajpal**
Follow up to https://github.com/code-dot-org/code-dot-org/pull/42243 with the second half of the levels.

A few other small changes in this PR:
- renames `this.effects` to `this.textEffects` in `PoemBotLibrary` to disambiguate from foreground effects.
- Doesn't fire line events in the first frame. This is because we show the whole poem in preview (frameCount == 1), which was accidentally firing all the line events immediately.
- Small update to the validation code for level 2 now that there can be multiple foreground effects at once (see https://github.com/code-dot-org/code-dot-org/pull/42259)

### CODE REVIEW

File: dashboard/config/scripts/levels/poetryHoC_textEffect.level [hunk 1]
```diff
@@ -39,7 +39,8 @@
     "preload_asset_list": null,
     "encrypted_examples": [
 
-    ]
+    ],
+    "validation_code": "var validationInfo = getValidationInfo();\r\nvar isFontSet = validationInfo.font.font != 'Arial';\r\nvar isColorSet = validationInfo.font.fill != 'black' || validationInfo.font.stroke != 'white';\r\nvar hasTextEffect = validationInfo.textEffects.length > 0;\r\n\r\n// Pass the level if they have at least two of the three effects.\r\nif (isFontSet && isColorSet || isFontSet && hasTextEffect || isColorSet && hasTextEffect) {\r\n  setSuccessFrame();\r\n}\r\ndrawProgressBar();\r\nif (World.frameCount == validationInfo.endTime) {\r\n  if (validationInfo.successFrame) {\r\n    if (isFontSet && isColorSet && hasTextEffect) {\r\n      // 3 effects\r\n      levelFailure(0, 'genericBonusSuccess');\r\n    } else {\r\n      // 2 effects\r\n      levelFailure(0, 'genericSuccess');\r\n    }\r\n  } else {\r\n    if (isFontSet || isColorSet || hasTextEffect) {\r\n      // 1 effect\r\n      levelFailure(3, 'poemBotOneTextEffect');\r\n    } else {\r\n      // 0 effects\r\n      levelFailure(3, 'poemBotNoTextEffects');\r\n    }\r\n  }\r\n}"
```
Comment (ajpal) [inline dashboard/config/scripts/levels/poetryHoC_textEffect.level:43]: image

File: dashboard/config/scripts/levels/poetryHoC_spriteBehaviors.level [hunk 2]
```diff
@@ -39,7 +39,8 @@
     "preload_asset_list": null,
     "encrypted_examples": [
 
-    ]
+    ],
+    "validation_code": "var spriteIds = getSpriteIdsInUse();\r\nvar validationInfo = getValidationInfo();\r\n\r\nif (!validationInfo.successFrame) {\r\n  for (var i = 0; i < spriteIds.length; i++) {\r\n    var behaviors = getBehaviorsForSpriteId(spriteIds[i]);\r\n    if (behaviors.length > 1 || (behaviors.length == 1 && behaviors[0] != 'fluttering')) {\r\n      setSuccessFrame();\r\n    }\r\n  }\r\n}\r\n\r\ndrawProgressBar();\r\nif (World.frameCount == validationInfo.endTime) {\r\n  if (validationInfo.successFrame) {\r\n    levelFailure(0, 'genericSuccess');\r\n  } else {\r\n    if (spriteIds.length == 0) {\r\n      levelFailure(3, 'poemBotDeletedSprite');\r\n    } else {\r\n      levelFailure(3, 'noBehavior');\r\n    }\r\n  }\r\n}"
```
Comment (ajpal) [inline dashboard/config/scripts/levels/poetryHoC_spriteBehaviors.level:43]: image

File: dashboard/config/scripts/levels/poetryHoC_events_text.level [hunk 3]
```diff
@@ -40,7 +40,8 @@
     "preload_asset_list": null,
     "encrypted_examples": [
 
-    ]
+    ],
+    "validation_code": "var spriteIds = getSpriteIdsInUse();\r\nvar validationInfo = getValidationInfo();\r\nif (!validationProps.previous) {\r\n  validationProps.previous = {};\r\n}\r\n\r\nvar hasLineEvent = validationInfo.lineEvents.length > 0;\r\nvar newTextEffect = validationProps.previous.numTextEffects < validationInfo.textEffects.length;\r\nif (hasLineEvent && newTextEffect) {\r\n  setSuccessFrame();\r\n}\r\nvalidationProps.previous.numTextEffects = validationInfo.textEffects.length;\r\n\r\ndrawProgressBar();\r\nif (World.frameCount == validationInfo.endTime) {\r\n  if (validationInfo.successFrame) {\r\n    levelFailure(0, 'genericSuccess');\r\n  } else {\r\n    levelFailure(3, 'poemBotTextEffectWithEvent');\r\n  }\r\n}"
```
Comment (ajpal) [inline dashboard/config/scripts/levels/poetryHoC_events_text.level:44]: image

File: dashboard/config/scripts/levels/poetryHoC_events1.level [hunk 4]
```diff
@@ -40,7 +40,8 @@
...
...us.backgroundEffect !=\r\n    validationInfo.backgroundEffect;\r\nif (hasLineEvent && (newForegroundEffect || newBackgroundEffect)) {\r\n  setSuccessFrame();\r\n  validationProps.previous.newForegroundEffect =\r\n    validationProps.previous.newForegroundEffect || newForegroundEffect;\r\n  validationProps.previous.newBackgroundEffect =\r\n    validationProps.previous.newBackgroundEffect || newBackgroundEffect;\r\n}\r\nvalidationProps.previous.numForegroundEffects = validationInfo.foregroundEffects.length;\r\nvalidationProps.previous.backgroundEffect = validationInfo.backgroundEffect;\r\n\r\ndrawProgressBar();\r\n\r\nif (World.frameCount == validationInfo.endTime) {\r\n  if (validationInfo.successFrame) {\r\n    if (validationProps.previous.newForegroundEffect && validationProps.previous.newBackgroundEffect) {\r\n      levelFailure(0, 'genericSuccess');\r\n    } else if (validationProps.previous.newForegroundEffect) {\r\n      levelFailure(0, 'poemBotSuccessForegroundEvent');\r\n    } else if (validationProps.previous.newBackgroundEffect) {\r\n      levelFailure(0, 'poemBotSuccessBackgroundEvent');\r\n    }\r\n  } else {\r\n    levelFailure(3, 'poemBotEffectWithEvent');\r\n  }\r\n}"
```
Comment (ajpal) [inline dashboard/config/scripts/levels/poetryHoC_events1.level:44]: image

File: dashboard/config/scripts/levels/poetryHoC_events_sprites.level [hunk 5]
```diff
@@ -40,7 +40,8 @@
...
...nvar addedSprite = false;\r\nfor (var i = 0; i < spriteIds.length; i++) {\r\n  if (!arrayIncludes(previousSpriteIds, spriteIds[i])) {\r\n    addedSprite = true;\r\n  }\r\n}\r\n\r\nvar removedSprite = false;\r\nfor (var i = 0; i < previousSpriteIds.length; i++) {\r\n  if (!arrayIncludes(spriteIds, previousSpriteIds[i])) {\r\n    removedSprite = true;\r\n  }\r\n}\r\n\r\nif (hasLineEvent && (addedSprite || removedSprite)) {\r\n  setSuccessFrame();\r\n  validationProps.addedSprite = validationProps.addedSprite || addedSprite;\r\n  validationProps.removedSprite = validationProps.removedSprite || removedSprite;\r\n}\r\nvalidationProps.previous.spriteIds = spriteIds;\r\n\r\ndrawProgressBar();\r\nif (World.frameCount == validationInfo.endTime) {\r\n  if (validationInfo.successFrame) {\r\n    if (validationProps.addedSprite && validationProps.removedSprite) {\r\n      levelFailure(0, 'genericSuccess');\r\n    } else if (validationProps.addedSprite) {\r\n      levelFailure(0, 'poemBotSuccessAddedSprite');\r\n    } else if (validationProps.removedSprite) {\r\n      levelFailure(0, 'poemBotSuccessRemovedSprite');\r\n    }\r\n  } else {\r\n    levelFailure(3, 'poemBotSpriteEvent');\r\n  }\r\n}"
```
Comment (ajpal) [inline dashboard/config/scripts/levels/poetryHoC_events_sprites.level:44]: image

**Review by maddiedierker — APPROVED** [review]

**Pull request merged**